File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []

        n = len(nums)
        nums.sort()

        for i, num in enumerate(nums):
            if num > 0:
                break
            
            if i > 0 and num == nums[i - 1]:
                continue

            l = i + 1
            r = n - 1
            
            while l < r and r < n:
                if nums[l] + nums[r] < -num:
                    l += 1
                elif nums[l] + nums[r] > -num:
                    r -= 1
                else:
                    ans.append([num, nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
        
        return ans

            

        

        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        std::vector<int> prefix(nums.size(), 1);
        std::vector<int> suffix(nums.size(), 1);
        std::vector<int> ans(nums.size(), 0);

        for (int i = 1; i < prefix.size(); i++) {
            prefix[i] = prefix[i-1] * nums[i-1]; 
        }

        for (int i = suffix.size() - 2; i >= 0; i--) {
            suffix[i] = suffix[i+1] * nums[i+1]; 
        }

        for (int i = 0; i < ans.size(); i++) {
            ans[i] = prefix[i] * suffix[i];
        }

        return ans;
    }
};
